module/VuFind/tests/unit-tests/src/VuFindTest/Service/ZoteroTest.php

```diff
@@ -0,0 +1,269 @@
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Psr7\Response;
+use Laminas\Cache\Storage\Adapter\AdapterOptions;
+use Laminas\Cache\Storage\StorageInterface;
+use Laminas\Session\Container;
+use League\OAuth1\Client\Credentials\CredentialsException;
+use League\OAuth1\Client\Credentials\TemporaryCredentials;
+use PHPUnit\Framework\MockObject\MockObject;
+use VuFind\Db\Entity\AccessToken;
+use VuFind\Db\Entity\User;
+use VuFind\Db\Service\AccessTokenService;
+use VuFind\Export\Zotero\ZoteroOAuth;
+use VuFind\Export\Zotero\ZoteroService;
+use VuFind\Export\Zotero\ZoteroTokenCredentials;
+use VuFind\Http\GuzzleService;
+use VuFindTest\Feature\FixtureTrait;
+
+/**
+ * Zotero Service Test Class
+ *
+ * @category VuFind
+ * @package  Tests
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:testing:unit_tests Wiki
+ */
+class ZoteroTest extends \PHPUnit\Framework\TestCase
+{
+    use FixtureTrait;
+
+    /**
+     * OAuth params
+     *
+     * @var array
+     */
+    protected array $oauthParams = [
+        'oauth_token' => 'token',
+        'oauth_verifier' => 'verifier',
+    ];
+
+    /**
+     * Zotero export URL
+     *
+     * @var string
+     */
+    protected string $zoteroExportUrl = 'https://api.zotero.org/users/zotero_user_id/items';
+
+    /**
+     * Cached access token
+     *
+     * @var ?AccessToken
+     */
+    protected ?AccessToken $storedAccessToken = null;
+
+    /**
+     * Container id for session storage
+     *
+     * @var int
+     */
+    protected static int $containerId = 0;
+
+    /**
+     * Test Zotero export.
+     *
+     * @return void
+     */
+    public function testExport(): void
+    {
+        $zotero = $this->getZoteroService(true);
+        $user = $this->getUser();
+        // First export goes to authorization:
+        $this->assertEquals('https://localhost/authorization', $zotero->export($user, 'https://localhost/callback'));
+        // Emulate return from Zotero authorization:
+        $this->assertEquals(1, $zotero->handleAuthCallback($user, $this->oauthParams));
+        // Subsequent call does export directly:
+        $this->assertEquals(1, $zotero->export($user, 'https://localhost/callback'));
+    }
+
+    /**
+     * Test authorization denied.
+     *
+     * @return void
+     */
+    public function testAuthorizationDenied(): void
+    {
+        $zotero = $this->getZoteroService(false, true);
+        $user = $this->getUser();
+        // First export goes to authorization:
+        $this->assertEquals('https://localhost/authorization', $zotero->export($user, 'https://localhost/callback'));
+        // Check that emulated return from Zotero authorization would throw:
+        $this->expectExceptionMessage('An error has occurred');
+        $this->assertEquals('https://localhost/authorization', $zotero->handleAuthCallback($user, $this->oauthParams));
```

Comment:
Does it make sense to have an assertion here? If the call throws an exception, the assertion will never be met anyway...


```suggestion
        $zotero->handleAuthCallback($user, $this->oauthParams);
```